### PR TITLE
Simplify if conditions

### PR DESCRIPTION
**Pull request by hermanventer**
Simplify if conditions before using them as the join condition for the true and else branches.

Beef up simplifier to simplify && and || expressions and to try to simplify equality expressions of the form "(cond ? xx : yy) (==|!=|===|!==) (undefined|null)" to just "cond" or "!cond"

### CODE REVIEW

File: src/utils/simplifier.js [hunk 1]
```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
...
+      return simplifyEquality(realm, value);
     default:
       return value;
   }
 }
 
+function simplifyEquality(realm: Realm, equality: AbstractValue): Value {
+  let op = equality.kind;
+  let [x, y] = equality.args;
+  if (y instanceof AbstractValue) [x, y] = [y, x];
```
Comment (NTillmann) [inline src/utils/simplifier.js]: You could reduce unneeded swapping by testing for `!(x instanceof AbstractValue)`, or `x instanceof ConcreteValue`.

File: src/utils/simplifier.js [hunk 2]
```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
   switch (value.kind) {
     case "!":
       return negate(realm, value.args[0]);
+    case "||":
+    case "&&":
+      let x = simplify(realm, value.args[0]);
```
Comment (yinghuitan) [inline src/utils/simplifier.js]: invariant(value.args.length === 2)

File: src/utils/simplifier.js [hunk 3]
```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
...
+function simplifyEquality(realm: Realm, equality: AbstractValue): Value {
+  let op = equality.kind;
+  let [x, y] = equality.args;
+  if (y instanceof AbstractValue) [x, y] = [y, x];
+  if (x instanceof AbstractValue && x.kind === "conditional" && (!y.mightNotBeUndefined() || !y.mightNotBeNull())) {
+    // try to simplify "(cond ? xx : yy) op undefined/null" to just "cond" or "!cond"
+    let [cond, xx, xy] = x.args;
+    invariant(cond instanceof AbstractValue); // otherwise the the conditional should not have been created
+    if (op === "===" || op === "!==") {
+      // if xx === undefined && yz !== undefined then cond <=> x === undefined
```
Comment (yinghuitan) [inline src/utils/simplifier.js]: All the comments use "yz" may need to change to "xy" in the code.

File: src/utils/simplifier.js (hunk 2 above)
Comment (hermanventer) [reply src/utils/simplifier.js]: This gets a bit tedious.

File: src/utils/simplifier.js (hunk 3 above)
Comment (hermanventer) [reply src/utils/simplifier.js]: Good catch. Will change.

File: src/utils/simplifier.js (hunk 1 above)
Comment (hermanventer) [reply src/utils/simplifier.js]: Fair enough. Will change.

File: src/utils/simplifier.js [hunk 4]
```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
...
       return value;
   }
 }
 
+function simplifyEquality(realm: Realm, equality: AbstractValue): Value {
+  let op = equality.kind;
+  let [x, y] = equality.args;
+  if (y instanceof AbstractValue) [x, y] = [y, x];
+  if (x instanceof AbstractValue && x.kind === "conditional" && (!y.mightNotBeUndefined() || !y.mightNotBeNull())) {
+    // try to simplify "(cond ? xx : yy) op undefined/null" to just "cond" or "!cond"
```
Comment (yinghuitan) [inline src/utils/simplifier.js]: "(cond ? xx : yy)" => "(cond ? xx : xy)" to align with code.

**Comment by hermanventer** [comment]
Requested change has been made.

File: src/utils/simplifier.js [hunk 5]
```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
...
+  let [x, y] = equality.args;
+  if (x instanceof ConcreteValue) [x, y] = [y, x];
+  if (x instanceof AbstractValue && x.kind === "conditional" && (!y.mightNotBeUndefined() || !y.mightNotBeNull())) {
+    // try to simplify "(cond ? xx : xy) op undefined/null" to just "cond" or "!cond"
+    let [cond, xx, xy] = x.args;
+    invariant(cond instanceof AbstractValue); // otherwise the the conditional should not have been created
+    if (op === "===" || op === "!==") {
+      // if xx === undefined && xy !== undefined then cond <=> x === undefined
+      if (!y.mightNotBeUndefined() && !xx.mightNotBeUndefined() && !xy.mightBeUndefined())
+        return op === "===" ? cond : negate(realm, cond);
```
Comment (NTillmann) [inline src/utils/simplifier.js]: Change requested: The original equality was guaranteed to yield a Boolean (true or false). This possibly replaces it with just `cond`. Is `cond` really guaranteed to be a Boolean? 

Prepacking the following doesn't seem to force a "Booleanization" of a numeric condition.

```js
  let x = __abstract("number", 42);
  y = x ? 1 : 2;
```

Similarly, I'd hope that `negate` always yields a Boolean in this setting, but I am not sure. Would be nice to see invariants for all this, especially around simplification rules.

File: src/utils/simplifier.js [hunk 6]
```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
...
+      if (!y.mightNotBeUndefined() && !xx.mightBeUndefined() && !xy.mightNotBeUndefined())
+        return op === "===" ? negate(realm, cond) : cond;
+      // if xx === null && xy !== null then cond <=> x === null
+      if (!y.mightNotBeNull() && !xx.mightNotBeNull() && !xy.mightBeNull())
+        return op === "===" ? cond : negate(realm, cond);
+      // if xx !== null && xy === null then !cond <=> x === null
+      if (!y.mightNotBeNull() && !xx.mightBeNull() && !xy.mightNotBeNull())
+        return op === "===" ? negate(realm, cond) : cond;
+    } else {
+      invariant(op === "==" || op === "!=");
```
Comment (NTillmann) [inline src/utils/simplifier.js]: All the checks and this nested invariant seems to imply a general precondition for this function: op.kind in ["==", "!=", "===", "!=="]

File: src/utils/simplifier.js (hunk 5 above)
Comment (hermanventer) [reply src/utils/simplifier.js]: Great catch. Will fix.

File: src/utils/simplifier.js (hunk 6 above)
Comment (hermanventer) [reply src/utils/simplifier.js]: It does, but I like having the derived invariant here instead.

**Comment by hermanventer** [comment]
Newly requested changes have been made.

File: src/utils/simplifier.js [hunk 7]
```diff
@@ -38,15 +39,68 @@ function simplify(realm, value: Value): Value {
...
+  invariant(value instanceof AbstractValue);
+  let v = AbstractValue.createFromUnaryOp(realm, "!", value, true, value.expressionLocation);
+  return AbstractValue.createFromUnaryOp(realm, "!", v, true, value.expressionLocation);
+}
+
 function negate(realm: Realm, value: Value): Value {
   if (value instanceof ConcreteValue) return ValuesDomain.computeUnary(realm, "!", value);
   invariant(value instanceof AbstractValue);
-  if (value.kind === "!") return value.args[0];
+  if (value.kind === "!") return makeBoolean(realm, value.args[0]);
```
Comment (NTillmann) [inline src/utils/simplifier.js]: Great! Would be nice if we had a better framework to state assumptions/guarantees. Right now, I had to chase through all code paths to verify the intended boolean-ness of the result. This doesn't scale when building more complex simplification rules.

Comment (hermanventer) [reply src/utils/simplifier.js]: There aren't any great ways of doing this that I can think of, but I'll get back to this when unifying all of the simplification code.

**Pull request closed without merging**